File: cpp/4.寻找两个正序数组的中位数.cpp
/*
 * @lc app=leetcode.cn id=4 lang=cpp
 *
 * [4] 寻找两个正序数组的中位数
 */

// @lc code=start
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int len1 = nums1.size(), len2 = nums2.size();
        int newlist[len1 + len2];
        for(int i = 0 ; i < len1; i++)
            newlist[i] = nums1[i];
        for(int i = 0 ; i < len2; i++)
            newlist[i + len1] = nums2[i];
        for(int i = 0 ; i < len1 + len2 - 1; i++)
            for(int j = 0 ; j + 1 < len1 + len2 - i; j++)
                if(newlist[j] > newlist[j + 1])
                {
                    int t = newlist[j];
                    newlist[j] = newlist[j + 1];
                    newlist[j + 1] = t;
                }        
        
        if( (len1+len2) % 2 == 0)
        {
            return double(newlist[(len1+len2)/2] + newlist[(len1+len2)/2 - 1])/2.0;
        }
        else
        {
            //cout << 2 << endl;
            return double(newlist[(len1+len2)/2]);
        }
    }
};
// @lc code=end

File: rs/4.寻找两个正序数组的中位数.rs
/*
 * @lc app=leetcode.cn id=4 lang=rust
 *
 * [4] 寻找两个正序数组的中位数
 */

// @lc code=start
impl Solution {
    pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
        let len1 = nums1.len();
        let len2 = nums2.len();
        let mid1 : usize = (len1 + len2 - 1) / 2;
        let mid2 : usize = (len1 + len2) / 2;
        let mut arr = [0,0];
        let mut sign = true;
        let mut i = 0;
        let mut j = 0;
        while(i + j <= mid2)
        {
            sign = !sign;
            if i == len1
            {
                if sign {
                    arr[1] = nums2[j];
                } else {
                    arr[0] = nums2[j];
                }
                j+=1;
                continue;
            }
            if j == len2
            {
                if sign {
                    arr[1] = nums1[i];
                } else {
                    arr[0] = nums1[i];
                }
                i+=1;
                continue;
            }
            if nums1[i] < nums2[j]
            {
                if sign {
                    arr[1] = nums1[i];
                } else {
                    arr[0] = nums1[i];
                }
                i+=1;
            }
            else
            {
                if sign {
                    arr[1] = nums2[j];
                } else {
                    arr[0] = nums2[j];
                }
                j+=1;
            }
        }

        if mid1 == mid2  {
            if sign 
            {
                arr[1] as f64
            }
            else
            {
                arr[0] as f64
            }
        } else
        {
            (arr[0] + arr[1]) as f64/2.0
        } 
    }
}
// @lc code=end

